File: app/lib/models/session.dart
class Session {
  final String id;
  final String organizationId;
  final String reservationId;
  final String coachId;
  final String memberId;
  final String? memberPackageId;
  final DateTime date;
  final String attendance;
  final String? memo;
  final dynamic workoutRecords;
  final String? feedback;
  final String? memberName;
  final String? coachName;
  final String? startTime;
  final String? endTime;

  const Session({
    required this.id,
    required this.organizationId,
    required this.reservationId,
    required this.coachId,
    required this.memberId,
    this.memberPackageId,
    required this.date,
    required this.attendance,
    this.memo,
    this.workoutRecords,
    this.feedback,
    this.memberName,
    this.coachName,
    this.startTime,
    this.endTime,
  });

  factory Session.fromJson(Map<String, dynamic> json) {
    return Session(
      id: json['id'] as String,
      organizationId: json['organizationId'] as String,
      reservationId: json['reservationId'] as String,
      coachId: json['coachId'] as String,
      memberId: json['memberId'] as String,
      memberPackageId: json['memberPackageId'] as String?,
      date: DateTime.parse(json['date'] as String),
      attendance: json['attendance'] as String? ?? 'PRESENT',
      memo: json['memo'] as String?,
      workoutRecords: json['workoutRecords'],
      feedback: json['feedback'] as String?,
      memberName: json['member']?['name'] as String?,
      coachName: json['coach']?['name'] as String?,
      startTime: json['reservation']?['startTime'] as String?,
      endTime: json['reservation']?['endTime'] as String?,
    );
  }
}

File: server/src/routes/session.ts
import { Router, Request, Response } from 'express';
import { z } from 'zod';
import { prisma } from '../utils/prisma';
import { authMiddleware } from '../middleware/auth';
import { parseDateOnly } from '../utils/kst-date';
import { requireCurrentOrgId, respondValidationError } from './_shared';

const router = Router();
router.use(authMiddleware);

// ─── List Sessions ─────────────────────────────────────────
router.get('/', async (req: Request, res: Response) => {
  try {
    const orgId = await requireCurrentOrgId(req, res);
    if (!orgId) return;

    const memberId = req.query.memberId as string | undefined;
    const startDate = req.query.startDate as string | undefined;
    const endDate = req.query.endDate as string | undefined;

    const where: any = { organizationId: orgId };
    if (memberId) where.memberId = memberId;
    if (startDate && endDate) {
      where.date = { gte: parseDateOnly(startDate), lte: parseDateOnly(endDate) };
    }

    const sessions = await prisma.session.findMany({
      where,
      include: {
        member: { select: { id: true, name: true } },
        coach: { select: { id: true, name: true } },
        reservation: { select: { startTime: true, endTime: true } },
      },
      orderBy: { date: 'desc' },
    });

    res.json(sessions);
  } catch (err) {
    console.error('List sessions error:', err);
    res.status(500).json({ error: 'Internal server error' });
  }
});

// ─── Get Session ───────────────────────────────────────────
router.get('/:id', async (req: Request, res: Response) => {
  try {
    const orgId = await requireCurrentOrgId(req, res);
    if (!orgId) return;

    const session = await prisma.session.findFirst({
      where: { id: req.params.id as string, organizationId: orgId },
      include: {
        member: { select: { id: true, name: true } },
        coach: { select: { id: true, name: true } },
        reservation: true,
        memberPackage: { include: { package: true } },
      },
    });

    if (!session) { res.status(404).json({ error: 'Session not found' }); return; }
    res.json(session);
  } catch (err) {
    console.error('Get session error:', err);
    res.status(500).json({ error: 'Internal server error' });
  }
});

// ─── Update Session ────────────────────────────────────────
const updateSessionSchema = z.object({
  memo: z.string().optional(),
  workoutRecords: z.any().optional(),
  feedback: z.string().optional(),
  attendance: z.enum(['PRESENT', 'NO_SHOW', 'LATE', 'CANCELLED']).optional(),
});

router.put('/:id', async (req: Request, res: Response) => {
  try {
    const orgId = await requireCurrentOrgId(req, res);
    if (!orgId) return;

    const body = updateSessionSchema.parse(req.body);

    const existingSession = await prisma.session.findFirst({
      where: { id: req.params.id as string, organizationId: orgId },
      select: { id: true },
    });
    if (!existingSession) {
      res.status(404).json({ error: 'Session not found' });
      return;
    }

    const session = await prisma.session.update({
      where: { id: existingSession.id },
      data: body,
    });

    res.json(session);
  } catch (err) {
    if (respondValidationError(res, err)) return;
    console.error('Update session error:', err);
    res.status(500).json({ error: 'Internal server error' });
  }
});

export default router;
